app/src/main/java/org/mozilla/focus/browser/FirefoxProgressBar.kt

```diff
@@ -26,23 +25,34 @@ private const val HIDE_ANIMATION_DURATION_MILLIS = 250L
 private val HIDE_AFTER_MILLIS = TimeUnit.SECONDS.toMillis(3)
 
 class FirefoxProgressBar @JvmOverloads constructor(
-        context: Context, attrs: AttributeSet? = null, defStyle: Int = 0)
-    : LinearLayout(context, attrs, defStyle) {
+    context: Context,
+    attrs: AttributeSet? = null,
+    defStyle: Int = 0
+) : LinearLayout(context, attrs, defStyle), Session.Observer {
 
     private val hideHandler = FirefoxProgressBarHideHandler(this)
+    private var session: Session? = null
 
     fun initialize(browserFrag: BrowserFragment) {
-        browserFrag.session.url.observe(browserFrag, Observer { sessionUrl -> url.text = sessionUrl })
-        browserFrag.session.loading.observe(browserFrag, object : NonNullObserver<Boolean>() {
-            public override fun onValueChanged(value: Boolean) {
-                val isLoading = value
-                if (isLoading) {
-                    showBar()
-                } else {
-                    scheduleHideBar()
-                }
+        session = browserFrag.session.also {
+            it.register(observer = this, owner = browserFrag)
```

Comment:
Do lifecycleOwners work correctly with fragment's view hierarchies? e.g. A fragment creates a view hierarchy, removes the view hierarchy, and adds another one, what happens? Do we still receive events to the listeners of the invalid view hierarchy? Are those objects prevented from being GC'd?

Reply:
This works the same as the observe() on the LiveData in the code before: The observer stays registered until the lifecycle of the owner changes to DESTROYED. So it's at least not a regression.
Attaching observers to the "lifecycle" of a View is something that is so common that we added something for that too. We can attach it to FirefoxProgressBar instead of the Fragment. I'll update the PR.